Handle two finger zoom and double click behavior

thanhdang198 opened the pull request:
This pull request includes significant changes to migrate from Mapbox to MapLibre and introduces new functionality for map interaction. The most important changes are grouped by theme as follows:

Migration from Mapbox to MapLibre:

* [`README.md`](diffhunk://#diff-b335630551682c19a781afebcf4d07bf978fb1f8ac04c6bf87428ed5106870f5L49-R49): Updated code examples to use `MapLibreMapOptions` and `MapLibreMap` instead of `MapboxMapOptions` and `MapboxMap`. [[1]](diffhunk://#diff-b335630551682c19a781afebcf4d07bf978fb1f8ac04c6bf87428ed5106870f5L49-R49) [[2]](diffhunk://#diff-b335630551682c19a781afebcf4d07bf978fb1f8ac04c6bf87428ed5106870f5L126-R125) [[3]](diffhunk://#diff-b335630551682c19a781afebcf4d07bf978fb1f8ac04c6bf87428ed5106870f5L135-R134)
* [`app/src/main/java/nl/flitsmeister/maplibrecar/MainActivity.kt`](diffhunk://#diff-81243589811e7608ab34f1aa816f09a84b7cbdab4f2cf91032aee8cb28990a80L13-R27): Replaced Mapbox imports and instances with MapLibre equivalents. [[1]](diffhunk://#diff-81243589811e7608ab34f1aa816f09a84b7cbdab4f2cf91032aee8cb28990a80L13-R27) [[2]](diffhunk://#diff-81243589811e7608ab34f1aa816f09a84b7cbdab4f2cf91032aee8cb28990a80L37-R37) [[3]](diffhunk://#diff-81243589811e7608ab34f1aa816f09a84b7cbdab4f2cf91032aee8cb28990a80L46-R50) [[4]](diffhunk://#diff-81243589811e7608ab34f1aa816f09a84b7cbdab4f2cf91032aee8cb28990a80L60-R60)
* [`car_common/src/main/java/nl/flitsmeister/car_common/CarMapContainer.kt`](diffhunk://#diff-e9c91cc61afecdf8cb57859c758ef10259291a50648d24be55c2dacbccf1e612R3-R24): Replaced Mapbox imports and instances with MapLibre equivalents. [[1]](diffhunk://#diff-e9c91cc61afecdf8cb57859c758ef10259291a50648d24be55c2dacbccf1e612R3-R24) [[2]](diffhunk://#diff-e9c91cc61afecdf8cb57859c758ef10259291a50648d24be55c2dacbccf1e612L31-R104) [[3]](diffhunk://#diff-e9c91cc61afecdf8cb57859c758ef10259291a50648d24be55c2dacbccf1e612L55-R122) [[4]](diffhunk://#diff-e9c91cc61afecdf8cb57859c758ef10259291a50648d24be55c2dacbccf1e612L68-R139) [[5]](diffhunk://#diff-e9c91cc61afecdf8cb57859c758ef10259291a50648d24be55c2dacbccf1e612L88-R156) [[6]](diffhunk://#diff-e9c91cc61afecdf8cb57859c758ef10259291a50648d24be55c2dacbccf1e612L100-R168) [[7]](diffhunk://#diff-e9c91cc61afecdf8cb57859c758ef10259291a50648d24be55c2dacbccf1e612R179)
* [`car_common/src/main/java/nl/flitsmeister/car_common/CarMapRenderer.kt`](diffhunk://#diff-52e2d1570a5988cff256b2fb02f8251fa5896968383174bdfc89a716bead6624L20-R22): Replaced Mapbox imports with MapLibre equivalents.

New functionality for map interaction:

* [`car_common/src/main/java/nl/flitsmeister/car_common/CarMapContainer.kt`](diffhunk://#diff-e9c91cc61afecdf8cb57859c758ef10259291a50648d24be55c2dacbccf1e612L31-R104): Added new methods for handling map scaling and double-click zoom with animation. [[1]](diffhunk://#diff-e9c91cc61afecdf8cb57859c758ef10259291a50648d24be55c2dacbccf1e612L31-R104) [[2]](diffhunk://#diff-e9c91cc61afecdf8cb57859c758ef10259291a50648d24be55c2dacbccf1e612L55-R122)
* [`car_common/src/main/java/nl/flitsmeister/car_common/CarMapRenderer.kt`](diffhunk://#diff-52e2d1570a5988cff256b2fb02f8251fa5896968383174bdfc89a716bead6624L154-R155): Simplified the `onScale` method by delegating to `CarMapContainer` and removed the custom zoom gesture generation. [[1]](diffhunk://#diff-52e2d1570a5988cff256b2fb02f8251fa5896968383174bdfc89a716bead6624L154-R155) [[2]](diffhunk://#diff-52e2d1570a5988cff256b2fb02f8251fa5896968383174bdfc89a716bead6624L190-L324)

Library version updates:

* [`gradle/libs.versions.toml`](diffhunk://#diff-697f70cdd88ba88fe77eebda60c7e143f6ad1286bca75017421e93ad84fb87dfL14-R14): Updated `mapLibreSdk` version from `10.2.0` to `11.7.0`.

Update contributor recogization

thanhdt-vietmap committed:
feat: upgrade maplibre map to `v11.7.0`

thanhdt-vietmap committed:
feat: integrate double click and two-finger zoom behavior

thanhdt-vietmap committed:
docs: update documentation and add contributor recognition

thanhdang198 committed:
Merge branch 'flitsmeister:main' into main

thanhdang198 commented:
Hi @frankkienl, I've created a PR for handle double click, two-finger zoom behavior as I said yet. 
Let me know if this PR has any problems. 
Thanks.

frankkienl commented on car_common/src/main/java/nl/flitsmeister/car_common/CarMapContainer.kt:
```diff
@@ -28,15 +34,76 @@ class CarMapContainer(
+                1.0,
+                zoomFocalPoint
+            )
+            scaleAnimator!!.start()
```
```suggestion
            scaleAnimator?.start()
```

thanhdang198 committed:
Update car_common/src/main/java/nl/flitsmeister/car_common/CarMapContainer.kt

Fix: Prevent potential NullPointerException by using safe call on scaleAnimator startup

Co-authored-by: Frank Bouwens <frankkie12345@gmail.com>

frankkienl commented:
It seems the zoom buttons don't work properly in this PR, but I found a solution for this; 
We can use the same animation used for the double-tap, for the zoom buttons too.

I've made a separate branch to demonstrate this solution:

// use the same animation for zoom in, and zoom out
https://github.com/flitsmeister/MapLibre-Android-Auto-Sample/blob/574079fe3adb8511172f95362848837473fd826f/car_common/src/main/java/nl/flitsmeister/car_common/CarMapContainer.kt#L71

// change the behavior of the zoom buttons, to act like the double-click
https://github.com/flitsmeister/MapLibre-Android-Auto-Sample/blob/574079fe3adb8511172f95362848837473fd826f/car_common/src/main/java/nl/flitsmeister/car_common/CarMapRenderer.kt#L144

If you're okay with that solution, we can merge this PR :-)

thanhdang198 commented:
> It seems the zoom buttons don't work properly in this PR, but I found a solution for this; We can use the same animation used for the double-tap, for the zoom buttons too.
> 
> I've made a separate branch to demonstrate this solution:
> 
> // use the same animation for zoom in, and zoom out
> 
> https://github.com/flitsmeister/MapLibre-Android-Auto-Sample/blob/574079fe3adb8511172f95362848837473fd826f/car_common/src/main/java/nl/flitsmeister/car_common/CarMapContainer.kt#L71
> 
> // change the behavior of the zoom buttons, to act like the double-click
> 
> https://github.com/flitsmeister/MapLibre-Android-Auto-Sample/blob/574079fe3adb8511172f95362848837473fd826f/car_common/src/main/java/nl/flitsmeister/car_common/CarMapRenderer.kt#L144
> 
> If you're okay with that solution, we can merge this PR :-)

There seems to be no problem with your solution.
